pycarol/carolina.py

```diff
@@ -25,49 +28,62 @@ def __init__(self, carol):
         self.cds_staging_rejected_path = None
 
     def init_if_needed(self):
+        """
+        Checks if backend credentials / object is still valid. This method is thread safe by using lock.
+
+        Returns:
+            True if backend has been reinitialized, false if it is still valid.
+
+        """
         expired = False
-        if self.client:
-            # Check if the token is not expired for at least another minute.. we do a little margin to avoid time difference issues
-            if self.expires_at is None or datetime.utcnow() + timedelta(minutes=1) < self.expires_at:
-                return
-            else:
-                expired = True
 
-        if Carolina.token is None or expired or Carolina.token.get('tenant_name', '') != self.carol.tenant['mdmName'] \
-                or Carolina.token.get('app_name', '') != self.carol.app_name \
-                or (datetime.utcnow() + timedelta(minutes=1)
-                    >= datetime.fromtimestamp(Carolina.token.get('expirationTimestamp', 1)/1000.0)):
-            token = self.carol.call_api('v1/storage/storage/token', params={'carolAppName': self.carol.app_name})
-            token['tenant_name'] = self.carol.tenant['mdmName']
-            token['app_name'] = self.carol.app_name
-            Carolina.token = token
+        # locks all instance variables to prevent inconsistencies
+        with Carolina.backend_lock:
+            if self.client:
+                # Check if the token is not expired for at least another minute.. we do a little margin to avoid time difference issues
+                if self.expires_at is None or datetime.utcnow() + timedelta(minutes=1) < self.expires_at:
+                    return False
+                else:
+                    expired = True
```

Comment:
this part does not need to be inside the thread lock.

Reply:
This first `if` is the key. If two threads evaluates it concurrently in a fraction of miliseconds they could both fall into `expired = True`. In that case they would start reseting the token, engine and all the instance vars, which might cause inconsistencies. 
By the code I guess both threads would set the exact same values, but I prefer leaving the whole method as a critical zone. In that way the first thread to get the lock would reset the vars, the second would wait, evatuate `expired = False` and just leave the method without any change.

Reply:
I did some tests and I was able to get a scenario that this could happen, but it is a very odd one. let's keep like this.